### PR TITLE
Fix map_top_n to use keys to break ties on NULL values. 

### LINKED ISSUE #22778: Improve map_top_n to consistently break ties for null values
## Expected Behavior
map_top_n should use the keys to compare when values are null , and break ties deterministically For e.g:
```
SELECT
    MAP_TOP_N(MAP(ARRAY[1, 2, 3, 4, 5], ARRAY[NULL, NULL, 1, 4, NULL]), 3);

{
  "5": null,
  "3": 1,
  "4": 4
}
```

## Current Behavior
Currently the output is indeterministic due to depending on map order:
```
SELECT
    MAP_TOP_N(MAP(ARRAY[1, 2, 3, 4, 5], ARRAY[NULL, NULL, 1, 4, NULL]), 3);

{
  "1": null,
  "3": 1,
  "4": 4
}
```

## Possible Solution
Fix map_top_n to fall back to comparing keys when values are null.

## Context
This is required for Presto/Velox equivalence.

### PR DESCRIPTION
**Pull request by kgpai**
## Description
Fix map_top_n to use keys to break ties for NULL values.

## Motivation and Context
Map_top_n currently uses keys to break ties when values match. This is however not done when values are null , which leads to some non determinism when compared agains the presto_native implementation.

fixes #22778

## Impact
Since current behavior is non deterministic, this has no impact on existing queries.

## Test Plan
See attached unit test.

## Release Notes
```
== RELEASE NOTES ==
General Changes
* Make map_top_n function more deterministic by using keys to break ties.
```

### CODE REVIEW

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/sql/TestMapTopNFunction.java [hunk 1]
```diff
@@ -95,6 +98,25 @@ public void testEmpty()
     public void testNull()
     {
         assertFunction("MAP_TOP_N(NULL, 1)", mapType(UNKNOWN, UNKNOWN), null);
+
+        // If values are null , then use keys to break ties.
```
Comment (elharo) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/sql/TestMapTopNFunction.java:102]: no space before comma

**Review by elharo — CHANGES_REQUESTED** [review]

**Comment by rschlussel** [comment]
Please update the documentation as well.

**Comment by kgpai** [comment]
@elharo Current behavior would depend on order of values in the valueBlock - I presume this means its dependent on order of input, and thus might be non deterministic. I will update and call out the present change to make it deterministic in the documentation.

**Comment by kgpai** [comment]
cc: @elharo @rschlussel Please have a look when possible. Thanks !

File: presto-docs/src/main/sphinx/functions/map.rst [hunk 2]
```diff
@@ -111,7 +111,7 @@ Map Functions
 
 .. function:: map_top_n_keys(x(K,V), n) -> array(K)
 
-    Returns top ``n`` keys in the map ``x`` by sorting its keys in descending order.
+    Returns top ``n`` keys in the map ``x`` by sorting its keys in descending order. Keys are used to break ties with the max key being chosen. Both keys and values should be orderable.
```
Comment (rschlussel) [inline presto-docs/src/main/sphinx/functions/map.rst:114]: I think you updated the wrong documentation.  This should be at map_top_n (a few entries below), not map_top_n_keys

Comment (kgpai) [reply presto-docs/src/main/sphinx/functions/map.rst]: Apologies and Thank you @rschlussel . Fixed.

File: presto-docs/src/main/sphinx/functions/map.rst [hunk 3]
```diff
@@ -136,7 +136,7 @@ Map Functions
 
 .. function:: map_top_n(x(K,V), n) -> map(K, V)
 
-    Truncates map items. Keeps only the top N elements by value.
+    Truncates map items. Keeps only the top N elements by value.  Keys are used to break ties with the max key being chosen. Both keys and values should be orderable.
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/functions/map.rst:139]:
```suggestion
    Truncates map items. Keeps only the top ``n`` elements by value.  Keys are used to break ties with the max key being chosen. Both keys and values should be orderable.
```
Nit, formatting.

File: presto-docs/src/main/sphinx/functions/map.rst [hunk 4]
```diff
@@ -136,7 +136,7 @@ Map Functions
...
     ``n`` must be a non-negative integer.::
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/functions/map.rst:140]:
```suggestion
    ``n`` must be a non-negative integer. ::
```
Nit: no space between the . and the first : results in "a non-negative integer.: " as shown in this screenshot 
Screenshot 2024-06-18 at 10 56 21 AM

**Review by steveburnett — CHANGES_REQUESTED** [review]
Thanks for the doc update! Two small nits, nothing big.

**Comment by kgpai** [comment]
@steveburnett Thank you, made the changes you requested.

File: presto-docs/src/main/sphinx/functions/map.rst [hunk 5]
```diff
@@ -136,8 +136,8 @@ Map Functions
 
 .. function:: map_top_n(x(K,V), n) -> map(K, V)
 
-    Truncates map items. Keeps only the top N elements by value.
-    ``n`` must be a non-negative integer.::
+    Truncates map items. Keeps only the top ``n``` elements by value.  Keys are used to break ties with the max key being chosen. Both keys and values should be orderable.
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/functions/map.rst:139]:
```suggestion
    Truncates map items. Keeps only the top ``n`` elements by value.  Keys are used to break ties with the max key being chosen. Both keys and values should be orderable.
```
Nit in formatting, nothing else.

**Review by steveburnett — CHANGES_REQUESTED** [review]
One final nit in formatting, nothing else!

**Comment by kgpai** [comment]
@steveburnett Thank you again ! Made changes as per your request.

**Review by rschlussel — APPROVED** [review]

**Review by steveburnett — APPROVED** [review]
LGTM! (docs)

Pull updated branch, reviewed new local docs build, looks good. Thanks!

**Comment by rschlussel** [comment]
@elharo can you review please?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/sql/MapSqlFunctions.java [hunk 6]
```diff
@@ -54,7 +54,8 @@ public static String mapKeysExists()
     @SqlType("map(K, V)")
     public static String mapTopN()
     {
-        return "RETURN IF(n < 0, fail('n must be greater than or equal to 0'), map_from_entries(slice(array_sort(map_entries(map_filter(input, (k, v) -> v is not null)), (x, y) -> IF(x[2] < y[2], 1, IF(x[2] = y[2], IF(x[1] < y[1], 1, -1), -1))) || map_entries(map_filter(input, (k, v) -> v is null)), 1, n)))";
+        return "RETURN IF(n < 0, fail('n must be greater than or equal to 0'), map_from_entries(slice(array_sort(map_entries(map_filter(input, (k, v) -> v is not null)), (x, y) -> IF(x[2] < y[2], 1, IF(x[2] = y[2], IF(x[1] < y[1], 1, -1), -1))) "
```
Comment (elharo) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/sql/MapSqlFunctions.java:57]: Not required but if there's any plausible way to use indentation and line breaks here to make this code easier to read, that would help a lot. Without that, it's very hard to follow the logic.

**Review by elharo — APPROVED** [review]
